96.md

```diff
@@ -0,0 +1,221 @@
+NIP-96
+======
+
+HTTP File Storage Integration
+-----------------------------
+
+`draft` `optional` `author:arthurfranca` `author:Semisol` `author:staab`
+
+## Introduction
+
+This NIP defines a REST API for HTTP file storage servers intended to be used in conjunction with the nostr network.
+The API will enable nostr users to upload files and later reference them by url on nostr notes.
+
+The spec DOES NOT use regular nostr events through websockets for
+storing, requesting nor retrieving data because, for simplicity, the server
+will not have to learn anything about nostr relays.
+
+## Server Adaptation
+
+File storage servers wishing to be accessible by nostr users should opt-in by making available an https route at `/.well-known/nostr/nip96.json` with `api_url`:
+
+```js
+{
+  // Required
+  // File upload and deletion are served from this url
+  // Also downloads if "download_url" field is absent or empty string
+  "api_url": "https://your-file-server.example/custom-api-path",
+  // Optional
+  // If absent, downloads are served from the api_url
+  "download_url": "https://a-cdn.example/a-path",
+  // Optional
+  // Use this if you are a nostr relay using your /.well-known/nostr/nip96.json
+  // just to advertise someone else's http file storage server's /.well-known/nostr/nip96.json
+  // In this case, "api_url" field must be an empty string
+  "recommended_url": "https://your-file-server.example",
+  // Optional
+  "supported_nips": [],
+  // Optional
+  "tos_url": "https://your-file-server.example/terms-of-service",
+  // Optional
+  "plans": {
+    // "free" is the only standardized plan key and
+    // clients may use its presence to learn if server offers free storage
+    "free": {
+      "name": "Free Tier",
+      "url": "https://...", // payment etc
+      "max_byte_size": 10485760,
+      // Range in days / 0 for no expiration
+      // [7, 0] means it may vary from 7 days to unlimited persistence,
+      // [0, 0] means it has no expiration
+      // early expiration may be due to low traffic or any other factor
+      "file_expiration": [14, 90],
+      "image_transformations": ["resizing"]
+    }
+  }
+}
+```
+
+Note that if using a CDN, the `server` MAY issue a 302 redirect from
```

Comment:
I want to start saying that this part was just advice, totally optional.

It talks about api when the server use the simple setup of using the same API url with GET for downloads. Or else it talks about dedicated download url.

It only says that, if for example you are charging for downloads (not uploads), and want to use NIP-98 for auth, CDNs can't be used directly because they don't know about NIP-98 nor even know that nostr is a thing.

Reply:
Nevermind, I saw you comment below and got that CDNs can understand the NIP-98 custom Authorization header. I will remove this paragraph.